frontend/packages/core/src/Contexts/workflow-storage-context/helpers.tsx

```diff
@@ -0,0 +1,56 @@
+import type { clutch as IClutch } from "@clutch-sh/api";
+
+import type { HydratedData } from "./types";
+
+// API data comes in as an array, this will rotate it into an object usable by the StorageContext
+const transformAPISharedState = (data: IClutch.shortlink.v1.IShareableState[]): HydratedData => {
+  const hydrated: HydratedData = {};
+
+  data.forEach(({ key, state = {} }) => {
+    hydrated[key] = state as any;
+  });
+
+  return hydrated;
+};
+
+const storeLocalData = (key: string, data: unknown) => {
+  try {
+    window.localStorage.setItem(key, JSON.stringify(data));
+  } catch (e) {
+    // eslint-disable-next-line no-console
+    console.error("Error saving to local storage", e);
+  }
+};
+
+const removeLocalData = (key: string) => window.localStorage.removeItem(key);
+
+const retrieveLocalData = (key: string) => {
+  const localData = window.localStorage.getItem(key);
+
+  if (localData) {
+    try {
+      return JSON.parse(localData);
+    } catch (_) {
+      return localData;
+    }
+  }
+};
+
+const retrieveData = (
+  store: HydratedData,
+  componentName: string,
+  key: string,
+  defaultData?: unknown
+): unknown => {
```

Comment:
Thought about this more after reviewing #2101 and realized we should probably be using generics here to avoid having to typecast.

Reply:
I've updated to use generics, tested it against the ProjectSelector and it seemed to work. If you dont mind giving it another pass and seeing if there's anything you'd like me to change about it.